source/aws_http_credentials_provider.c

```diff
@@ -0,0 +1,515 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/auth/credentials.h>
+#include <aws/auth/private/aws_http_credentials_provider.h>
+#include <aws/auth/private/credentials_utils.h>
+#include <aws/common/clock.h>
+#include <aws/http/request_response.h>
+#include <aws/http/status_code.h>
+#include <aws/io/channel_bootstrap.h>
+#include <aws/io/socket.h>
+#include <aws/io/tls_channel_handler.h>
+
+#define HTTP_RESPONSE_SIZE_INITIAL 2048
+#define HTTP_RESPONSE_SIZE_LIMIT 10000
+#define HTTP_CONNECT_TIMEOUT_DEFAULT_IN_SECONDS 2
+#define HTTP_MAX_ATTEMPTS 3
+#define HTTP_RETRY_TIMEOUT_MS 100
+
+struct aws_credentials_provider_http_impl {
+    struct aws_http_connection_manager *connection_manager;
+    const struct aws_auth_http_system_vtable *function_table;
+    struct aws_string *endpoint;
+    struct aws_retry_strategy *retry_strategy;
+    struct aws_http_credentials_provider_request_vtable *request_vtable;
+};
+
+static void s_http_query_context_reset_request_specific_data(struct aws_http_query_context *http_query_context) {
+    struct aws_credentials_provider_http_impl *impl = http_query_context->provider->impl;
+    impl->request_vtable->destroy_request_data_fn(http_query_context->request_data);
+    if (http_query_context->request) {
+        aws_http_message_release(http_query_context->request);
+        http_query_context->request = NULL;
+    }
+    if (http_query_context->connection) {
+        int result = impl->function_table->aws_http_connection_manager_release_connection(
+            impl->connection_manager, http_query_context->connection);
+        (void)result;
+        AWS_ASSERT(result == AWS_OP_SUCCESS);
+        http_query_context->connection = NULL;
+    }
+    http_query_context->status_code = 0;
+    http_query_context->error_code = 0;
+    aws_byte_buf_clean_up(&http_query_context->path_and_query);
+}
+
+static void s_http_query_context_destroy(struct aws_http_query_context *http_query_context) {
+    if (http_query_context == NULL) {
+        return;
+    }
+
+    s_http_query_context_reset_request_specific_data(http_query_context);
+    aws_mem_release(http_query_context->allocator, http_query_context->request_data);
+    aws_byte_buf_clean_up(&http_query_context->payload);
```

Comment:
We should move this into the `s_http_query_context_reset_request_specific_data`. On failure, there could also be some payload in the response received. And for the retry, we should have clean up the previous received payload as well.